dattri/algorithm/influence_function.py

```diff
@@ -222,169 +225,362 @@ def attribute(
+                data_target_pair[data_item_idx] = torch.cat(
+                    (
+                        data_target_pair[data_item_idx],
+                        data_target_pair_list[batch_idx][data_item_idx],
+                    ),
+                    dim=0,
+                )
+
+        # to device (only once for all data)
+        for data_item_idx in range(len(data_target_pair)):
+            data_target_pair[data_item_idx] = data_target_pair[data_item_idx].to(
+                self.device,
+            )
+
+        func = partial(self.task.get_loss_func(), data_target_pair=data_target_pair)
+
         from dattri.func.projection import arnoldi_project
 
-        if not hasattr(self, "arnoldi_projector"):
-            feature_dim = query.shape[1]
-            func = partial(self.task.get_loss_func(), data_target_pair=train_data)
-            model_params, _ = self.task.get_param(index)
-            self.arnoldi_projector = arnoldi_project(
-                feature_dim,
-                func,
-                model_params,
-                device=self.device,
-                **transformation_kwargs,
+        for i in range(len(self.task.get_checkpoints())):
+            model_params, _ = self.task.get_param(i)
+            self.arnoldi_projectors.append(
+                arnoldi_project(
+                    feature_dim=len(model_params),
+                    func=func,
+                    x=model_params,
+                    proj_dim=self.proj_dim,
+                    max_iter=self.max_iter,
+                    norm_constant=self.norm_constant,
+                    tol=self.tol,
+                    regularization=self.regularization,
+                    seed=self.seed,
+                    device=self.device,
+                ),
             )
 
-        return self.arnoldi_projector(query).detach()
+    def transform_test_rep(
+        self,
+        ckpt_idx: int,
+        test_rep: torch.Tensor,
+    ) -> torch.Tensor:
+        """Transform the test representations via Arnoldi projection.
+
+        Args:
+            ckpt_idx (int): The index of the model checkpoints. This index
+                is used for ensembling different trained model checkpoints.
+            test_rep (torch.Tensor): The test representations to be transformed.
+                It is a 2-d dimensional tensor with the shape of
+                (batch_size, num_params).
+
+        Returns:
+            torch.Tensor: The transformed test representations, a 2-d dimensional
+                tensor with the shape of (batch_size, proj_dim).
+
+        Raises:
+            ValueError: If the Arnoldi projector has not been cached.
+        """
+        if not hasattr(self, "arnoldi_projectors"):
+            error_msg = "The Arnoldi projector has not been cached.\
+                         Please call cache() first."
+            raise ValueError(error_msg)
+
+        return self.arnoldi_projectors[ckpt_idx](test_rep).detach()
```

Comment:
I think we also need to transform train_rep? @tingwl0122

Reply:
for the arnoldi projector written in `dattri/func/projection.py`,
the projection is defined as (eigvecs * inverse of eigvals * eigvecs.T) * (test or train feature),
so we only need to apply it to one of test or train;

while in the original paper, it define the projection as (eigvec.T) * (test or train feature),
so it needs to apply projection to both train and test, along with the inverse eigvals in the middle.

Reply:
For the original paper, the projection is defined as "eigvec.T * train/test feature";
while we defined it as "eigvec * inverse eigvals * eigvec.T * train/test feature".
Thus, we only need to apply it to either one of the train or test feature

Reply:
In this case, we are storing a num_parameter x num_parameter matrix in the projector?

It's probably more memory efficient to define the projection as (sqrt(1/eigvals) * eigvec.T)[topk_index]?

Reply:
I think this is because we regard this as a inverse-hessian multiplier before.

Reply:
Yeah, that's likely the reason. Now that we treat it as a projection, could you please raise a PR to make this more efficient?

Reply:
Just raised a PR for this